docs/getting-started/walkthroughs/datastream_api.md

```diff
@@ -0,0 +1,897 @@
+This way, when a large transaction comes through, you can check if the flag is set for that account.
+If it is, then this is fraud and output an alert.
+This flag is what you want to store in Flink state.
+
+The most basic type of state in Flink is [ValueState]({{ site.baseurl }}/dev/stream/state/state.html#using-managed-keyed-state), a data type that provides _fault tolerant_, _managed_, _per key_ state.
+`ValueState` is created using a `ValueStateDescriptor` which contains metadata about how it should be managed.
+
+<div class="codetabs" markdown="1">
+<div data-lang="java" markdown="1">
+{% highlight java %}
+public class FraudDetector extends KeyedProcessFunction<Long, Transaction, Alert> {
+
+	public static final double SMALL_AMOUNT = 0.01;
+
+	public static final double LARGE_AMOUNT = 500.00;
+
+	public static final long ONE_DAY = 24 * 60 * 60 * 1000;
+
+	private transient ValueState<Boolean> flagState;
+
+	@Override
+	public void open(Configuration parameters) throws Exception {
+		ValueStateDescriptor<Boolean> flagDescriptor = new ValueStateDescriptor<>(
+				"flag",
+				Types.BOOLEAN);
+
+		flagState = getRuntimeContext().getState(flagDescriptor);
+	}
+{% endhighlight %}
+</div>
+
+<div data-lang="scala" markdown="1">
+{% highlight scala %}
+class FraudDetector extends KeyedProcessFunction[Long, Transaction, Alert] {
+
+    @transient private var flagState: ValueState[java.lang.Boolean] = _
+
+    @throws[Exception]
+    override def open(parameters: Configuration): Unit = {
+        val flagDescriptor = new ValueStateDescriptor("flag", Types.BOOLEAN)
+        flagState = getRuntimeContext.getState(flagDescriptor)
+    }
+{% endhighlight %}
+</div>
+</div>
+
+`ValueState` is a wrapper class, similar to `AtomicReference` in the Java standard library.
+It provides three methods for interacting with its contents; `update` sets the state, `value` gets the current value, and `clear` to delete its contents.
+Otherwise, fault tolerance is managed automatically under the hood, and so you can interact with it like any standard variable.
+
+Below, you can see an example of how you can use a flag state to track potential fraudulent transactions.
+
+<div class="codetabs" markdown="1">
+<div data-lang="java" markdown="1">
+{% highlight java %}
+    @Override
+    public void processElement(
+        Transaction transaction,
+        Context context,
+        Collector<Alert> collector) throws Exception {
+
+        // Get the current state for the current key
+        Boolean lastTransactionWasSmall = flagState.value();
+
+        if (lastTransactionWasSmall != null) {
+            if (transaction.getAmount() > LARGE_AMOUNT) {
+                //Output an alert downstream
+                Alert alert = new Alert();
+                alert.setId(transaction.getAccountId());
+
+                collector.collect(alert);            
+            }
+        }
+        
+        // clean up our state
+        flagState.clear();
```

Comment:
How about putting this into the `lastTransactionWasSmall != null` statement right at the end so that we only call this if needed?